### PR TITLE
Adjust comment message applied to dataset detailed audit log entries

### PR DESCRIPTION
**Pull request by Sigmonia**
#### Rationale
Update audit log message for merged dataset rows that contain no changes

#### Changes
* Adjust messaging for merges applied to dataset detailed audit log entries

### CODE REVIEW

File: study/src/org/labkey/study/model/DatasetDefinition.java [hunk 1]
```diff
@@ -1662,7 +1662,8 @@ protected AuditTypeEvent createSummaryAuditRecord(User user, Container c, AuditC
         @Override
         protected DatasetAuditProvider.DatasetAuditEvent createDetailedAuditRecord(User user, Container c, AuditConfigurable tInfo, QueryService.AuditAction action, @Nullable String userComment, @Nullable Map<String, Object> record, Map<String, Object> existingRecord)
         {
-            String auditComment = null!=userComment ? userComment :
+            boolean userCommentIsBlank = StringUtils.trimToNull(userComment) == null;
+            String auditComment = !userCommentIsBlank ? userComment :
```
Comment (labkey-susanh) [inline study/src/org/labkey/study/model/DatasetDefinition.java]: I'm not sure this is really what's wanted here.  "userComment" was introduced for sampleManagement timeline and is meant to be text captured from a user when a change is made that likely documents why the change was made. Is there UI available for adding a userComment for datasets?  Or is the thought that this will possibly come in through an API call?

File: study/src/org/labkey/study/model/DatasetDefinition.java [hunk 2]
```diff
@@ -1686,16 +1680,31 @@ protected AuditTypeEvent createSummaryAuditRecord(User user, Container c, AuditC
...
                 Pair<Map<String, Object>, Map<String, Object>> rowPair = AuditHandler.getOldAndNewRecordForMerge(record, existingRecord, Collections.emptySet());
                 oldRecordString = DatasetAuditProvider.encodeForDataMap(c, rowPair.first);
                 newRecordString = DatasetAuditProvider.encodeForDataMap(c, rowPair.second);
+
+                // Check if no fields changed, if so adjust messaging
+                if (rowPair.second.size() == 0 && userCommentIsBlank)
+                {
+                    auditComment = "Dataset row was processed, but no changes detected";
+                    newRecordString = DatasetAuditProvider.encodeForDataMap(c, record);
+                }
```
Comment (labkey-susanh) [inline study/src/org/labkey/study/model/DatasetDefinition.java:1695]: You could make an `if-else` out of this and line 1687, but I still question the test for `userCommentIsBlank`.

File: study/src/org/labkey/study/model/DatasetDefinition.java (hunk 1 above)
Comment (Sigmonia) [reply study/src/org/labkey/study/model/DatasetDefinition.java]: userComment is coming from the AbstractAuditHandler.createDetailedAuditRecord API, while I agree in this case it is always null and probably not exposed. If/when it is implemented, then I think it would take precedence.

Maybe it would be better to create an AuditRecord builder class? Then it wouldn't need to be passed in here?

File: study/src/org/labkey/study/model/DatasetDefinition.java (hunk 2 above)
Comment (Sigmonia) [reply study/src/org/labkey/study/model/DatasetDefinition.java:1695]: fixed.

File: study/src/org/labkey/study/model/DatasetDefinition.java (hunk 1 above)
Comment (labkey-susanh) [reply study/src/org/labkey/study/model/DatasetDefinition.java]: We actually use a separate field for the UserComment because the "comment" field is meant to still record the more generic "A row was inserted" or "A row was updated" and the user comment is allow the user to say why it was done..  See InventoryDomainKind .  This userComment field was not added in every domain (so, not in AbstractAuditDomainKind), but I think it should be used similarly in other places when there is call for it.

A builder class could make sense, but I wouldn't do it in this PR.

**Review by labkey-susanh — CHANGES_REQUESTED** [review]

**Review by labkey-susanh — APPROVED** [review]
Thanks for adding the javadocs.
